src/modal/modal-stack.ts

```diff
@@ -20,75 +22,110 @@ import {NgbActiveModal, NgbModalRef} from './modal-ref';
-      this._applicationRef.attachView(backdropCmptRef.hostView);
-      containerEl.appendChild(backdropCmptRef.location.nativeElement);
-    }
-    windowCmptRef = this._windowFactory.create(this._injector, contentRef.nodes);
-    this._applicationRef.attachView(windowCmptRef.hostView);
-    containerEl.appendChild(windowCmptRef.location.nativeElement);
-
-    ngbModalRef = new NgbModalRef(windowCmptRef, contentRef, backdropCmptRef, options.beforeDismiss);
+    let backdropCmptRef: ComponentRef<NgbModalBackdrop> =
+        options.backdrop !== false ? this._attachBackdrop(containerEl) : null;
+    let windowCmptRef: ComponentRef<NgbModalWindow> = this._attachWindowComponent(containerEl, contentRef);
+    let ngbModalRef: NgbModalRef = new NgbModalRef(windowCmptRef, contentRef, backdropCmptRef, options.beforeDismiss);
 
     activeModal.close = (result: any) => { ngbModalRef.close(result); };
     activeModal.dismiss = (reason: any) => { ngbModalRef.dismiss(reason); };
 
     this._applyWindowOptions(windowCmptRef.instance, options);
-
     return ngbModalRef;
   }
 
+  // Cant build factories on constructor when using lazy loading
+  private _buildFactories() {
+    this._buildBackdropFactory();
+    this._buildWindowFactory();
+  }
+
+  private _buildBackdropFactory() {
+    if (this._backdropFactory == null) {
+      this._backdropFactory = this._componentFactoryResolver.resolveComponentFactory(NgbModalBackdrop);
+    }
+  }
+
+  private _buildWindowFactory() {
+    if (this._windowFactory == null) {
+      this._windowFactory = this._windowFactory =
+          this._componentFactoryResolver.resolveComponentFactory(NgbModalWindow);
+    }
+  }
+
+  private _attachBackdrop(containerEl: any): ComponentRef<NgbModalBackdrop> {
+    let backdropCmptRef = this._backdropFactory.create(this._injector);
+    this._applicationRef.attachView(backdropCmptRef.hostView);
+    containerEl.appendChild(backdropCmptRef.location.nativeElement);
+    return backdropCmptRef;
+  }
+
+  private _attachWindowComponent(containerEl: any, contentRef: any): ComponentRef<NgbModalWindow> {
+    let windowCmptRef = this._windowFactory.create(this._injector, contentRef.nodes);
+    this._applicationRef.attachView(windowCmptRef.hostView);
+    containerEl.appendChild(windowCmptRef.location.nativeElement);
+    return windowCmptRef;
+  }
+
   private _applyWindowOptions(windowInstance: NgbModalWindow, options: Object): void {
-    ['backdrop', 'keyboard', 'size', 'windowClass'].forEach((optionName: string) => {
+    this._windowAttributes.forEach((optionName: string) => {
       if (isDefined(options[optionName])) {
         windowInstance[optionName] = options[optionName];
       }
     });
   }
 
-  private _getContentRef(
-      moduleCFR: ComponentFactoryResolver, contentInjector: Injector, content: any,
-      context: NgbActiveModal): ContentRef {
+  private _getContentRef(contentInjector: Injector, content: any, context: NgbActiveModal): ContentRef {
```

Comment:
The signature should stay as-is, we need to use a proper `ComponentFactoryResolver`